### PR TITLE
[Merged by Bors] - feat: port NumberTheory.BernoulliPolynomials

### PR DESCRIPTION
**Pull request by mo271**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mo271** [comment]
`apply_congr` not working as good as it used to...

**Comment by adomani** [comment]
The only issues left in this PR appear to be due to `apply_congr` no longer working as in Lean 3: should the broken proofs be fixed, avoiding `apply_congr` or should `apply_congr` be fixed instead?

**Comment by mo271** [comment]
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/!4.234352.20apply_congr.20in.20BernoulliPolynomials

**Comment by mo271** [comment]
Great fix @alexjbest! I think this is ready to go...

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

File: Mathlib/NumberTheory/BernoulliPolynomials.lean [hunk 1]
```diff
@@ -90,63 +91,59 @@ theorem bernoulli_eval_zero (n : ℕ) : (bernoulli n).eval 0 = bernoulli n := by
...
   rw [mul_assoc, mul_assoc, ← Nat.cast_mul, ← Nat.cast_mul]
   congr 3
   rw [(choose_mul_succ_eq k m).symm, mul_comm]
 #align polynomial.derivative_bernoulli_add_one Polynomial.derivative_bernoulli_add_one
 
-theorem derivative_bernoulli (k : ℕ) : (bernoulli k).derivative = k * bernoulli (k - 1) := by
-  cases k
-  · rw [Nat.cast_zero, MulZeroClass.zero_mul, bernoulli_zero, derivative_one]
-  · exact_mod_cast derivative_bernoulli_add_one k
-#align polynomial.derivative_bernoulli Polynomial.derivative_bernoulli
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/BernoulliPolynomials.lean]: Why would you remove this entire theorem?

Comment (alexjbest) [reply Mathlib/NumberTheory/BernoulliPolynomials.lean]: Probably just a mistake when trying some different edits out. If you want to add it back go ahead, if not I'll try and do it later today

**Pull request closed without merging**